File: android/src/main/kotlin/net/spross/app/ui/DatesReference.kt
package net.spross.app.ui

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import net.spross.app.AppModel
import net.spross.app.Chrome
import net.spross.app.dateSprosse
import net.spross.kern.catalog.DateDrillContent
import net.spross.kern.catalog.dateNotes
import net.spross.kern.trainer.DateDrill
import net.spross.kern.trainer.DateReferenceRow

/**
 * The reading half of the dates page: the seven weekdays and the twelve months, both sides
 * beside each other.
 *
 * The table is [DateDrill.reference] — the same joined rows the run grades against, so it
 * cannot claim one name and ask for another. Under a learned name stand the forms the drill
 * also accepts and teaches: its short form, its other lexemes (de `Sonnabend`), and what it
 * becomes inside a date where that differs (uk `березня`). How a sheet is read is
 * [ReferenceSection].
 *
 * The generated table can say nothing the rows do not carry, so how the language ASSEMBLES
 * a date — and what trips a learner up doing it — is authored prose under it
 * (`catalog/dates/<lang>.json` § dateNotes), the numbers page's own shape.
 */
@Composable
fun DateReferenceSection(model: AppModel, content: DateDrillContent, chrome: Chrome) {
    val groups = remember(content, chrome) {
        DateDrill.reference(content).map { group ->
            ReferenceGroup(
                // One bare-name pool — the Sprosse rows above already name the two.
                caption = chrome.dateSprosse(listOf(group.kind)),
                rows = group.rows.map {
                    ReferenceRow(source = it.source, target = it.target, targetUnder = otherForms(it))
                },
            )
        }
    }
    ReferenceSection(model, chrome.datesReference, groups, content.source, content.target, chrome)

    val notes = remember(content) { model.catalog?.dateNotes(content.target, content.source).orEmpty() }
    if (notes.isNotEmpty()) {
        OverviewHeading(chrome.commonNotes)
        OverviewPanel {
            for (note in notes) {
                Text(note, style = MaterialTheme.typography.bodyMedium)
            }
        }
    }
}

/** What else the learned name answers to; null keeps the caption line off the row. */
private fun otherForms(row: DateReferenceRow): String? {
    val forms = listOfNotNull(row.abbr) + row.synonyms + listOfNotNull(row.dateForm)
    return if (forms.isEmpty()) null else forms.joinToString(" · ")
}

File: App/Sources/Screens/DatesReference.swift
import SwiftUI
import SprossKern

/// The reading half of the dates overview: the seven weekdays and the twelve
/// months, both sides beside each other. The page around it is `DrillOverview`,
/// which asks the face for this section and knows nothing of what is in it.
///
/// The table is `DateDrill.reference` — the same joined rows the run grades
/// against, so it cannot claim one name and ask for another. Under a learned
/// name stand the forms the drill also accepts and teaches: its short form,
/// its other lexemes (de `Sonnabend`), and what it becomes inside a date where
/// that differs (uk `березня`).
///
/// What the rows cannot carry — how the language ASSEMBLES a date — is the
/// authored prose under them (`catalog/dates/<lang>.json` § dateNotes).
struct DatesReference: View {
    let model: AppModel
    let content: DateDrillContent
    /// The language the learner KNOWS — the left of every row.
    let source: String
    /// The language being learned — the right of it, and the side a tap says.
    let target: String

    var body: some View {
        VStack(alignment: .leading, spacing: Theme.spacing.lg) {
            ReferenceSheet(heading: "dates.reference",
                           groups: DateDrill.shared.reference(content: content).map {
                               // The Sprosse rows above already name the two pools, so the
                               // group headings reuse their words rather than authoring a
                               // second pair — the face's table, never a second copy of the
                               // numbering, which is what drifted the day the ladder grew a
                               // Sprosse.
                               ReferenceGroup(title: DateDrillFace.sprosseTitle([$0.kind]), rows: $0.rows)
                           },
                           speak: { speak($0.target) },
                           row: nameRow)
            ReferenceNotes(lines: model.catalog?.dateNotes(language: target,
                                                           reader: model.sourceLanguage) ?? [])
        }
    }

    /// One name, twice: the known language on the left, the learned one on the
    /// right, with the learned side's other forms under it.
    ///
    /// A tap says the LEARNED side only: the other column is the reader's own
    /// language, and a reference sheet is read to hear what one cannot yet say.
    private func nameRow(_ row: DateReferenceRow) -> some View {
        HStack(alignment: .top, spacing: Theme.spacing.md) {
            Text(verbatim: row.source)
                .font(Theme.typography.headline)
                .foregroundStyle(Theme.colors.textPrimary)
                .spoken(row.source, language: source)
            Spacer(minLength: Theme.spacing.sm)
            VStack(alignment: .trailing, spacing: 2) {
                Text(verbatim: row.target)
                    .font(Theme.typography.headline)
                    .foregroundStyle(Theme.colors.accent)
                    .spoken(row.target, language: target)
                if let under = Self.otherForms(row) {
                    Text(verbatim: under)
                        .font(Theme.typography.caption)
                        .foregroundStyle(Theme.colors.textSecondary)
                        .spoken(under, language: target)
                }
            }
            .multilineTextAlignment(.trailing)
        }
        .fixedSize(horizontal: false, vertical: true)
    }

    /// What else the learned name answers to, on one caption line: the short
    /// form the prompt wears, the other lexemes, and the in-a-date form.
    private static func otherForms(_ row: DateReferenceRow) -> String? {
        var forms: [String] = []
        if let abbr = row.abbr { forms.append(abbr) }
        forms += row.teaches
        if let dateForm = row.dateForm { forms.append(dateForm) }
        return forms.isEmpty ? nil : forms.joined(separator: " · ")
    }

    /// Hearing a name in the language being learned — nil where the device can
    /// neither play nor say it, so the page offers no sound it cannot make.
    private func speak(_ name: String) -> (() -> Void)? {
        model.pronounceAction(for: name, lang: target)
    }
}
